File: typescript/src/nodes.ts
/**
 * Tenant nodes resource — list, current, set-default.
 *
 * Backed by the Infinity control plane (the URL passed as `infinityURL`),
 * NOT the active tenant node. Each node corresponds to one regional
 * control plane (e.g. node1.prodxcloud.com).
 */

import type { Transport } from './transport.js';
import type { NodeInfo } from './types.js';

export class Nodes {
  constructor(private t: Transport) {}

  async list(): Promise<NodeInfo[]> {
    const res = await this.t.get<{ nodes?: NodeInfo[] } | NodeInfo[]>('/api/v1/auth/nodes/');
    if (Array.isArray(res.body)) return res.body.map(wrap);
    return ((res.body as { nodes?: NodeInfo[] })?.nodes ?? []).map(wrap);
  }

  async current(): Promise<NodeInfo | null> {
    const list = await this.list();
    return list.find((n) => n.isDefault) ?? list[0] ?? null;
  }

  async setDefault(idOrName: string | number): Promise<{ success: boolean; node?: NodeInfo }> {
    const res = await this.t.postJSON<{ success: boolean; node?: NodeInfo }>(
      '/api/v1/auth/nodes/default',
      { id: idOrName },
    );
    return res.body ?? { success: false };
  }

  /** Register a self-hosted vxnode container (BYO hardware). Mirrors the
   *  dashboard's SelfHostedNodeForm. POST /api/v1/auth/nodes/self-hosted. */
  async registerSelfHosted(input: {
    hostname: string;
    customDomainName: string;
    port?: number;
    publicIp?: string;
    privateIp?: string;
    tunnelProvider?: 'cloudflare' | 'tailscale' | 'caddy' | 'direct';
    keyPairName?: string;
    ideConnectionToken?: string;
    agent1ConnectionToken?: string;
    storageType?: 'ssd' | 'hdd' | 'nvme';
    storageBackupMode?: 'none' | 'daily' | 'weekly';
    description?: string;
    sshUsername?: string;
  }): Promise<NodeInfo> {
    if (!input.hostname || !input.customDomainName) {
      throw new Error('nodes.registerSelfHosted: hostname + customDomainName required');
    }
    const body: Record<string, unknown> = {
      hostname: input.hostname,
      custom_domain_name: input.customDomainName.replace(/^https?:\/\//, '').replace(/\/+$/, ''),
      port: input.port ?? 8744,
      tunnel_provider: input.tunnelProvider ?? 'cloudflare',
      storage_type: input.storageType ?? 'ssd',
      storage_backup_mode: input.storageBackupMode ?? 'none',
    };
    if (input.publicIp) body.public_ip = input.publicIp;
    if (input.privateIp) body.private_ip = input.privateIp;
    if (input.keyPairName) body.key_pair_name = input.keyPairName;
    if (input.ideConnectionToken) body.ide_connection_token = input.ideConnectionToken;
    if (input.agent1ConnectionToken) body.agent1_connection_token = input.agent1ConnectionToken;
    if (input.description) body.description = input.description;
    if (input.sshUsername) body.ssh_username = input.sshUsername;
    const res = await this.t.postJSON<NodeInfo>(
      '/api/v1/auth/nodes/self-hosted', body,
    );
    return wrap(res.body ?? {});
  }

  /** Delete a node record. DELETE /api/v1/auth/nodes/{id}. The caller is
   *  responsible for terminating any underlying VM first. */
  async delete(id: string | number): Promise<{ status: string; message?: string }> {
    if (!id && id !== 0) throw new Error('nodes.delete: id is required');
    const res = await this.t.delete<{ status: string; message?: string }>(
      `/api/v1/auth/nodes/${encodeURIComponent(String(id))}`,
    );
    return res.body ?? { status: 'ok' };
  }

  /** Partial-update an existing node. PATCH /api/v1/auth/nodes/{id}. Only
   *  fields you set are sent. Backend rejects `publicIp` / `instanceId` —
   *  those are platform-managed. */
  async update(id: string | number, patch: {
    hostname?: string;
    customDomainName?: string;
    loadBalancer?: string;
    privateIp?: string;
    status?: 'active' | 'running' | 'stopped' | 'provisioning' | 'pending' | 'failed' | 'terminated';
    isDefaultNode?: boolean;
    providerComputeType?: string;
    storageType?: 'ssd' | 'hdd' | 'nvme';
    storageBackupMode?: 'none' | 'daily' | 'weekly';
    storageBackupAddress?: string;
    installationChecklist?: Array<Record<string, unknown>>;
    enabledFeatures?: unknown[];
    vpnAccessDetails?: Record<string, unknown>;
    tunnelVm?: Record<string, unknown>;
  }): Promise<NodeInfo> {
    if (!id && id !== 0) throw new Error('nodes.update: id is required');
    const body: Record<string, unknown> = {};
    if (patch.hostname !== undefined) body.hostname = patch.hostname;
    if (patch.customDomainName !== undefined) {
      body.custom_domain_name = patch.customDomainName.replace(/^https?:\/\//, '').replace(/\/+$/, '');
    }
    if (patch.loadBalancer !== undefined) body.load_balancer = patch.loadBalancer;
    if (patch.privateIp !== undefined) body.private_ip = patch.privateIp;
    if (patch.status !== undefined) body.status = patch.status;
    if (patch.isDefaultNode !== undefined) body.is_default_node = patch.isDefaultNode;
    if (patch.providerComputeType !== undefined) body.provider_compute_type = patch.providerComputeType;
    if (patch.storageType !== undefined) body.storage_type = patch.storageType;
    if (patch.storageBackupMode !== undefined) body.storage_backup_mode = patch.storageBackupMode;
    if (patch.storageBackupAddress !== undefined) body.storage_backup_address = patch.storageBackupAddress;
    if (patch.installationChecklist !== undefined) body.installation_checklist = patch.installationChecklist;
    if (patch.enabledFeatures !== undefined) body.enabled_features = patch.enabledFeatures;
    if (patch.vpnAccessDetails !== undefined) body.vpn_access_details = patch.vpnAccessDetails;
    if (patch.tunnelVm !== undefined) body.tunnel_vm = patch.tunnelVm;
    if (Object.keys(body).length === 0) {
      throw new Error('nodes.update: at least one field must be set');
    }
    const res = await this.t.patchJSON<NodeInfo>(
      `/api/v1/auth/nodes/${encodeURIComponent(String(id))}`, body,
    );
    return wrap(res.body ?? {});
  }
}

function wrap(n: NodeInfo | Record<string, unknown>): NodeInfo {
  const b = n as Record<string, unknown>;
  return {
    id: (b.id as string | number) ?? '',
    name: (b.name as string) ?? '',
    url: b.url as string | undefined,
    isDefault: Boolean(b.is_default ?? b.default),
    raw: b,
  };
}

File: nodes/nodes.go
// Package nodes is the resource module for tenant-node management.
//
// Tenant nodes are per-customer provisioning environments — each customer
// has at least one, and SDK calls target one specific node at a time.
//
// All endpoints in this module live on the Infinity control plane
// (api.vxcloud.io), not on a tenant node, because the node registry is
// a control-plane concern.
package nodes

import (
	"context"
	"strconv"
	"strings"

	"github.com/prodxcloud/vxcloud/transport"
)

// Node mirrors infinity's NodeResponse pydantic model.
type Node struct {
	ID               int     `json:"id"`
	UserID           int     `json:"user_id,omitempty"`
	OrganizationID   string  `json:"organization_id,omitempty"`
	Hostname         string  `json:"hostname"`
	Status           string  `json:"status"`
	IsDefaultNode    bool    `json:"is_default_node"`
	PublicIP         string  `json:"public_ip,omitempty"`
	PrivateIP        string  `json:"private_ip,omitempty"`
	InstanceID       string  `json:"instance_id,omitempty"`
	LoadBalancer     string  `json:"load_balancer,omitempty"`
	CustomDomainName string  `json:"custom_domain_name,omitempty"`
	OSType           string  `json:"os_type,omitempty"`
	BoxType          string  `json:"box_type,omitempty"`
	StorageGB        float64 `json:"storage_gb,omitempty"`
	Category         string  `json:"category,omitempty"`
	License          string  `json:"license,omitempty"`
	SessionID        string  `json:"session_id,omitempty"`
	StatePath        string  `json:"state_path,omitempty"`
	CreatedAt        string  `json:"created_at,omitempty"`
	UpdatedAt        string  `json:"updated_at,omitempty"`
}

// BaseURL returns the resolved https URL for the node, with the same
// priority order vxcli and the frontend use:
// custom_domain_name → load_balancer → public_ip.
//
// Returns "" if the node has no addressable hostname yet.
func (n Node) BaseURL() string {
	raw := strings.TrimSpace(n.CustomDomainName)
	if raw == "" {
		raw = strings.TrimSpace(n.LoadBalancer)
	}
	if raw == "" {
		raw = strings.TrimSpace(n.PublicIP)
	}
	if raw == "" {
		return ""
	}
	if !strings.HasPrefix(raw, "http://") && !strings.HasPrefix(raw, "https://") {
		raw = "https://" + raw
	}
	return strings.TrimRight(raw, "/")
}

// Label is a short user-facing identifier.
func (n Node) Label() string {
	switch {
	case n.CustomDomainName != "":
		return n.CustomDomainName
	case n.Hostname != "":
		return n.Hostname
	case n.LoadBalancer != "":
		return n.LoadBalancer
	default:
		return "node-" + strconv.Itoa(n.ID)
	}
}

// Client is the nodes resource module facade. Constructed by the parent
// vxsdk.Client; callers obtain it from c.Nodes().
type Client struct {
	T           *transport.Transport
	InfinityURL string
}

// List returns all tenant nodes registered to the authenticated principal.
//
// Endpoint: GET {InfinityURL}/api/v1/auth/nodes/.
//
// Note: vxcli's `node list` command returned 401 for some tokens during
// preview testing. The SDK's auto-refresh-on-401 handles that case
// transparently when an API key is configured on the Client.
func (c *Client) List(ctx context.Context) ([]Node, error) {
	u := transport.JoinURL(c.InfinityURL, "/api/v1/auth/nodes/")
	var out []Node
	if err := c.T.JSON(ctx, "nodes.List", "GET", u, nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// SetDefault marks the given node ID as the user's default. Subsequent
// API key exchanges will resolve this node as the tenant target.
//
// Endpoint: POST {InfinityURL}/api/v1/auth/nodes/{id}/set-default.
func (c *Client) SetDefault(ctx context.Context, id int) error {
	u := transport.JoinURL(c.InfinityURL, "/api/v1/auth/nodes/"+strconv.Itoa(id)+"/set-default")
	return c.T.JSON(ctx, "nodes.SetDefault", "POST", u, struct{}{}, nil)
}

// Default returns the node currently marked as the workspace default,
// or nil if none exists.
func (c *Client) Default(ctx context.Context) (*Node, error) {
	all, err := c.List(ctx)
	if err != nil {
		return nil, err
	}
	for i := range all {
		if all[i].IsDefaultNode {
			return &all[i], nil
		}
	}
	return nil, nil
}

// RegisterSelfHostedInput mirrors SelfHostedNodeRegisterRequest in
// app/auth/node.py:772 — the dashboard's "Self-Hosted" form payload.
type RegisterSelfHostedInput struct {
	Hostname              string `json:"hostname"`
	CustomDomainName      string `json:"custom_domain_name"`
	Port                  int    `json:"port,omitempty"`
	PublicIP              string `json:"public_ip,omitempty"`
	PrivateIP             string `json:"private_ip,omitempty"`
	TunnelProvider        string `json:"tunnel_provider,omitempty"`
	KeyPairName           string `json:"key_pair_name,omitempty"`
	IDEConnectionToken    string `json:"ide_connection_token,omitempty"`
	Agent1ConnectionToken string `json:"agent1_connection_token,omitempty"`
	StorageType           string `json:"storage_type,omitempty"`
	StorageBackupMode     string `json:"storage_backup_mode,omitempty"`
	Description           string `json:"description,omitempty"`
	SSHUsername           string `json:"ssh_username,omitempty"`
}

// RegisterSelfHosted registers a vxnode container the caller is running
// themselves (BYO hardware). No VM is provisioned on our side — the row
// is created with cloud_provider=self-hosted, status=active.
//
// Endpoint: POST {InfinityURL}/api/v1/auth/nodes/self-hosted.
func (c *Client) RegisterSelfHosted(ctx context.Context, in RegisterSelfHostedInput) (*Node, error) {
	u := transport.JoinURL(c.InfinityURL, "/api/v1/auth/nodes/self-hosted")
	var out Node
	if err := c.T.JSON(ctx, "nodes.RegisterSelfHosted", "POST", u, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a node record. Only deletes the row — the caller is
// responsible for terminating any underlying VM first.
//
// Endpoint: DELETE {InfinityURL}/api/v1/auth/nodes/{id}.
func (c *Client) Delete(ctx context.Context, id int) error {
	u := transport.JoinURL(c.InfinityURL, "/api/v1/auth/nodes/"+strconv.Itoa(id))
	return c.T.JSON(ctx, "nodes.Delete", "DELETE", u, nil, nil)
}

// UpdateInput mirrors NodeUpdateRequest in app/auth/node.py:81. All fields
// are pointer/slice/map so the zero value can be distinguished from "not
// set" — only non-nil fields are sent in the PATCH body. Backend rejects
// any field not in its EDITABLE_FIELDS whitelist (public_ip is read-only).
type UpdateInput struct {
	Hostname              *string                  `json:"hostname,omitempty"`
	CustomDomainName      *string                  `json:"custom_domain_name,omitempty"`
	LoadBalancer          *string                  `json:"load_balancer,omitempty"`
	PrivateIP             *string                  `json:"private_ip,omitempty"`
	Status                *string                  `json:"status,omitempty"`
	IsDefaultNode         *bool                    `json:"is_default_node,omitempty"`
	ProviderComputeType   *string                  `json:"provider_compute_type,omitempty"`
	StorageType           *string                  `json:"storage_type,omitempty"`
	StorageBackupMode     *string                  `json:"storage_backup_mode,omitempty"`
	StorageBackupAddress  *string                  `json:"storage_backup_address,omitempty"`
	InstallationChecklist []map[string]interface{} `json:"installation_checklist,omitempty"`
	EnabledFeatures       []interface{}            `json:"enabled_features,omitempty"`
	VPNAccessDetails      map[string]interface{}   `json:"vpn_access_details,omitempty"`
	TunnelVM              map[string]interface{}   `json:"tunnel_vm,omitempty"`
}

// Update partial-updates an existing node record. Only fields set in `in`
// are sent. Endpoint: PATCH {InfinityURL}/api/v1/auth/nodes/{id}.
func (c *Client) Update(ctx context.Context, id int, in UpdateInput) (*Node, error) {
	u := transport.JoinURL(c.InfinityURL, "/api/v1/auth/nodes/"+strconv.Itoa(id))
	var out Node
	if err := c.T.JSON(ctx, "nodes.Update", "PATCH", u, in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
